### PR TITLE
[Feature] Gesture recognition algorithm MTUT on NVGesture dataset

### PR DESCRIPTION
**Pull request by Ben-Louis**
## Motivation

Add hand gesture recognition function to mmpose.

## Modification

- add NVGesture dataset
- add I3D backbone to extract features from videos
- add MTUT head
- add GestureRecognizer to support general gesture recognition algorithms
- add configs for training/test of I3D+MTUT model on NVGesture dataset

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: tools/webcam/configs/gesture/try_det.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (jin-s13) [inline tools/webcam/configs/gesture/try_det.py:1]: Is this file really needed?

File: mmpose/models/heads/mtut_head.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        for i, modal in enumerate(img_metas['modality']):
+            out = self.avg_pool(x[i])
+            out = self.dropout(out)
+            out = getattr(self.output_conv, modal)(out)
+            out = out.mean(3).mean(3)
+            logits.append(out)
+        return logits
+
+    @staticmethod
+    def _compute_corr(fmap):
```
Comment (jin-s13) [inline mmpose/models/heads/mtut_head.py:71]: we need docstring for each major function

File: mmpose/models/heads/mtut_head.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        self._train_epoch = 0
+
+    def init_weights(self):
+        """Initialize model weights."""
+        for m in self.output_conv.modules():
+            if isinstance(m, nn.Conv3d):
+                xavier_init(m)
+
+    def set_train_epoch(self, epoch: int):
+        self._train_epoch = epoch
```
Comment (jin-s13) [inline mmpose/models/heads/mtut_head.py:58]: what is `_train_epoch` for?

File: tools/webcam/configs/gesture/try_det.py (hunk 1 above)
Comment (Ben-Louis) [reply tools/webcam/configs/gesture/try_det.py:1]: No. It will be replaced in the coming commit with a clean config.

File: mmpose/datasets/datasets/gesture/nvgesture_dataset.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            res_folder (str, optional): The folder to save the testing
+                results. If not specified, a temp folder will be created.
+                Default: None.
+            metric (str | list[str]): Metric to be performed.
+                Options: 'AP'.
+
+        Returns:
+            dict: Evaluation results for evaluation metric.
+        """
+        if metric != 'AP':
```
Comment (jin-s13) [inline mmpose/datasets/datasets/gesture/nvgesture_dataset.py:146]: The metric could be a list ['AP'], I will recommend the following way.
```
metrics = metric if isinstance(metric, list) else [metric]
allowed_metrics = ['AP']
for metric in metrics:
   if metric not in allowed_metrics:
        raise KeyError(f'metric {metric} is not supported')
```

Comment (jin-s13) [reply mmpose/datasets/datasets/gesture/nvgesture_dataset.py:146]: Please also double check AP or mAP?

Comment (Ben-Louis) [reply mmpose/datasets/datasets/gesture/nvgesture_dataset.py:146]: Here mAP represents the mean AP among different modalities. It can be used to select the best model. `AP_{modal}` represents the AP of a single modality.

File: configs/hand/gesture_sview_rgbd_vid/mtut/nvgesture/i3d_nvgesture_bbox_112x112_fps15_rgb.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+# learning policy
+lr_config = dict(policy='step', gamma=0.1, step=[30, 60, 90, 110])
+total_epochs = 130
+log_config = dict(interval=10)
+
+custom_hooks_config = [dict(type='ModelSetEpochHook')]
+
+model = dict(
+    type='GestureRecognizer',
+    modality=['rgb'],
```
Comment (jin-s13) [inline configs/hand/gesture_sview_rgbd_vid/mtut/nvgesture/i3d_nvgesture_bbox_112x112_fps15_rgb.py:25]: Is this config for rgb only? 
Then the config file path `gesture_sview_rgbd_vid`  is a little bit confusing.

File: mmpose/datasets/datasets/gesture/nvgesture_dataset.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                results. If not specified, a temp folder will be created.
+                Default: None.
+            metric (str | list[str]): Metric to be performed.
+                Options: 'AP'.
+
+        Returns:
+            dict: Evaluation results for evaluation metric.
+        """
+        metrics = metric if isinstance(metric, list) else [metric]
+        allowed_metrics = ['AP']
```
Comment (jin-s13) [inline mmpose/datasets/datasets/gesture/nvgesture_dataset.py:149]: AP or mAP?

Comment (Ben-Louis) [reply mmpose/datasets/datasets/gesture/nvgesture_dataset.py:149]: AP. Is it less confusing if I change 'mAP' to 'AP_mean'?

File: mmpose/models/backbones/utils/transformer.py [hunk 7]
```diff
@@ -1,1156 +0,0 @@
...
-            shape with input.
-    """
-    x = x.clamp(min=0, max=1)
-    x1 = x.clamp(min=eps)
-    x2 = (1 - x).clamp(min=eps)
-    return torch.log(x1 / x2)
-
-
-# @TRANSFORMER_LAYER.register_module()
-class DetrTransformerDecoderLayer(BaseTransformerLayer):
```
Comment (jin-s13) [inline mmpose/models/backbones/utils/transformer.py:404]: Are these class deleted?

File: mmpose/datasets/pipelines/gesture_transform.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+@PIPELINES.register_module()
+class TemporalPooling:
+    """Pick frames according to either stride or reference fps.
+
+    Required keys: 'video', 'modality', 'num_frames', 'fps'.
+
+    Modified keys: 'video', 'num_frames'.
+    """
+
+    def __init__(self, length: int = 64, stride: int = 1, ref_fps: int = -1):
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/gesture_transform.py:48]: add introduction to these args

File: configs/hand/gesture_sview_rgbd_vid/mtut/nvgesture/i3d_nvgesture_bbox_112x112_fps15_rgb.py (hunk 5 above)
Comment (Ben-Louis) [reply configs/hand/gesture_sview_rgbd_vid/mtut/nvgesture/i3d_nvgesture_bbox_112x112_fps15_rgb.py:25]: Yes. As stated in `/configs/hand/gesture_sview_rgbd_vid/mtut/nvgesture/i3d_nvgesture.md`, MTUT supports multi-modal training and uni-modal inference. This config is mainly used for demo and webcam with RGB videos.

File: mmpose/datasets/pipelines/gesture_transform.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+            results['num_frames'] = length
+        del results['valid_frames']
+        if 'bbox' in results:
+            results['bbox'] = results['bbox'][start:start + length]
+        return results
+
+
+@PIPELINES.register_module()
+class TemporalPooling:
+    """Pick frames according to either stride or reference fps.
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/gesture_transform.py:41]: Does this mean stride and reference fps should not be set simultaneously?

File: mmpose/datasets/pipelines/gesture_transform.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+                                                -1, num_frames)
+            results['video'][i] = results['video'][i].transpose(3, 0, 1, 2)
+            results['width'][i], results['height'][i] = video.shape[
+                1], video.shape[0]
+
+        return results
+
+
+@PIPELINES.register_module()
+class NVGestureRandomFlip:
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/gesture_transform.py:201]: It seems that this pipeline is specific to NVGesture dataset? Can we make it more general?

File: mmpose/datasets/pipelines/gesture_transform.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+        self.mean = torch.tensor(mean)
+        self.std = torch.tensor(std)
+
+    def __call__(self, results):
+        """Implement data normalization."""
+        for i, modal in enumerate(results['modality']):
+            if modal == 'rgb':
+                video = results['video'][i]
+                dim = video.ndim - 1
+                video = video.sub(self.mean.view(3, *((1, ) * dim)))
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/gesture_transform.py:397]: Can we simply use `-` instead of `sub` ?

File: mmpose/datasets/datasets/gesture/nvgesture_dataset.py (hunk 6 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/gesture/nvgesture_dataset.py:149]: AP_mean looks good.

**Review by jin-s13 — COMMENTED** [review]
Looks good to me in general. Unittests are required before merging.

File: mmpose/models/backbones/utils/transformer.py (hunk 7 above)
Comment (Ben-Louis) [reply mmpose/models/backbones/utils/transformer.py:404]: It is deleted in https://github.com/open-mmlab/mmpose/pull/1405.

File: mmpose/datasets/pipelines/gesture_transform.py (hunk 9 above)
Comment (Ben-Louis) [reply mmpose/datasets/pipelines/gesture_transform.py:41]: Yes, it seems better to add an assertion here.

File: mmpose/models/backbones/utils/transformer.py (hunk 7 above)
Comment (jin-s13) [reply mmpose/models/backbones/utils/transformer.py:404]: ok

File: mmpose/datasets/pipelines/gesture_transform.py (hunk 10 above)
Comment (Ben-Louis) [reply mmpose/datasets/pipelines/gesture_transform.py:201]: Ok. Horizontal flip sometimes changes the category of the input video. This pipeline needs to know the label pairs that represent lateral symmetric gestures. How about making such label pairs a parameter for this pipeline, or adding `flipped` and `label_after_flip` items in `results`?

Comment (jin-s13) [reply mmpose/datasets/pipelines/gesture_transform.py:201]: This is similar to https://github.com/open-mmlab/mmpose/blob/1e8f1e53cf8218fa575a1b2bdccdf52f9ac8adbb/mmpose/datasets/pipelines/top_down_transform.py#L129

File: mmpose/datasets/pipelines/gesture_transform.py [hunk 12]
```diff
@@ -0,0 +1,400 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import mmcv
+import numpy as np
+import torch
+
+from mmpose.core import bbox_xywh2xyxy, bbox_xyxy2xywh
+from mmpose.datasets.builder import PIPELINES
+
+
+@PIPELINES.register_module()
```
[the remark this answers is not in the document]
Comment (Ben-Louis) [reply mmpose/datasets/pipelines/gesture_transform.py:10]: Unittests of all gesture pipelines except VideoColorJitter have been added.

File: mmpose/datasets/pipelines/gesture_transform.py (hunk 8 above)
Comment (Ben-Louis) [reply mmpose/datasets/pipelines/gesture_transform.py]:
```
  Args:
      length (int): output video length. If unset, the entire video will
          be pooled.
      stride (int): temporal pooling stride. If unset, the stride will be
          computed with video fps and `ref_fps`. If both `stride` and
          `ref_fps` are unset, the stride will be 1.
      ref_fps (int): expected fps of output video. If unset, the video will
          be pooling with `stride`.
```

File: mmpose/datasets/pipelines/gesture_transform.py (hunk 11 above)
Comment (Ben-Louis) [reply mmpose/datasets/pipelines/gesture_transform.py]: Ok, both `sub` and `div` are replaced.

File: mmpose/datasets/pipelines/gesture_transform.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jin-s13) [reply mmpose/datasets/pipelines/gesture_transform.py:10]: ok

**Review by jin-s13 — APPROVED** [review]

**Comment by ly015** [comment]
@Ben-Louis CI failed.

**Review by ly015 — APPROVED** [review]
